### PR TITLE
g_assert should be an expression not a statement; provide g_assertf f is format like printf

### PR DESCRIPTION
**Pull request by jaykrell**
g_assert should be an expression not a statement; provide g_assertf f is format like printf

- g_assert should be an expression, not a statement
   ANSI C assert gets this correct, and I corrected
   widely used NT ASSERT circa 2001 with no ill effects
   Note that ANSI C assert is void, g_assert here is boolean.

 - provide g_assertf
   f is for format, like printf -- too terse?
   Where you might previously say:
   if (!(expr))
     g_error(...); // like printf
   now:
    g_assertf(expr, ...)

Manual tests provided -- how to provide a test whose success is failure to compile?
Correct existing eglib/test to build and pass, but don't enable in build.
Enabling it is a line in toplevel configure.ac to add eglib/test/Makefile
after eglib/Makefile.

### CODE REVIEW

**Comment by vargaz** [comment]
Why is this needed ?

**Comment by jaykrell** [comment]
The expression part is generally correct though I don't remember all why. I figured this out in 2001, fixed the widely used existing system and got it right ever since without thinking about it. In general, expressions fit in more places than statements and statements seldom/never are advantageous. Statements are sometimes forced upon you, and are often just an accident. Here the statement is probably just an accident -- like because people often don't know how to write expressions. I didn't look through history here. Perhaps it was deliberate. g_assertf is useful because Ludovic was having me use if (!expr) g_error where I had a perhaps underinformative g_assert(expr). I totally agree assert should be informative -- I often see people say if (x) assert(false) and I always tell them that assert(false) is bad. I went to make it easier by providing g_assertf and noticed g_assert poorly done.

**Comment by jaykrell** [comment]
build

**Comment by jaykrell** [comment]
The old assert is there for two reasons: easier to read the diff, easier to test. Once the change is successfully read by a human, could move it to the test, or "dismiss as obvious, heavily used, must be working, no need for tests" and remove the old and even the test. I've been way down the slippery slope of "obvious, heavily used, no need for testing" so am trying not to here -- worse, the test isn't automated -- let me see about moving eglib forward separately and we can revisit...

**Comment by jaykrell** [comment]
Statements are arguably a style. They prevent getting fancy -- am I allowed to intermingle ternary and assert or does that get difficult to read? Whether or not that policy belongs here, unclear, but probably not.

**Comment by jaykrell** [comment]
@monojenkins build Linux ARMv5 soft float

**Comment by jaykrell** [comment]
@monojenkins build Linux ARMv7 hard float

**Comment by jaykrell** [comment]
@monojenkins build Linux ARMv7 hard float

**Comment by jaykrell** [comment]
@monojenkins build Linux ARMv5 soft float

**Comment by jaykrell** [comment]
@monojenkins build Linux x64

**Comment by jaykrell** [comment]
@monojenkins build Linux x64

File: mono/eglib/glib.h [hunk 1]
```diff
@@ -719,9 +719,45 @@ GUnicodeBreakType   g_unichar_break_type (gunichar c);
...
+g_assertf is a boolean expression -- the precise value is not preserved, just true or false.
+
+Other than expr, the parameters are not evaluated unless expr is false.
+
+format must be a string literal, in order to be concatenated.
+If this is too restrictive, g_error remains.
+*/
+#define  g_assertf(x, format, ...) \
+	(G_LIKELY(x) ? 1 : (g_assertion_message ( \
+	"* Assertion at %s:%d, condition `%s' not met, function:%hs, " format "\n", \
```
Comment (lambdageek) [inline mono/eglib/glib.h]: `%hs` is not a valid format specifier in standard C.  If we want to support this MSVC non-standard extension, it should be ifdef'd.

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 FullAOT

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64

**Comment by jaykrell** [comment]
@monojenkins build Linux ARMv5 soft float

**Pull request merged**